### PR TITLE
[ADT] Backport std::to_underlying from C++23

### PR DESCRIPTION
**Pull request by Endilll**
This patch backports a one-liner `std::to_underlying` that came with C++23. This is useful for refactoring unscoped enums into scoped enums, because the latter are not implicitly convertible to integer types.

I followed libc++ implementation, but I consider their testing too heavy for us, so I wrote a simpler set of tests.

### CODE REVIEW

**Review by AaronBallman — APPROVED** [review]
The changes look reasonable to me, and I presume you intend to use this in-tree for the enum changes you've been working on in Clang. But I added a few more reviewers just to get more sets of eyes on the changes; please don't land until at least one more reviewer has accepted.

**Comment by Endilll** [comment]
Formatting is fine. I guess I pushed formatting commit too fast for bot to notice.

**Comment by slinder1** [comment]
The code LGTM, but with one small nit: as it is a backport of a C++23 feature please move it to `include/llvm/ADT/STLForwardCompat.h`

File: llvm/include/llvm/ADT/STLExtras.h [hunk 1]
```diff
@@ -1535,6 +1535,13 @@ constexpr decltype(auto) makeVisitor(CallableTs &&...Callables) {
   return detail::Visitor<CallableTs...>(std::forward<CallableTs>(Callables)...);
 }
 
+/// Backport of C++23 std::to_underlying.
+template <typename Enum>
+[[nodiscard]] constexpr typename std::underlying_type_t<Enum>
+to_underlying(Enum E) {
+  return static_cast<typename std::underlying_type_t<Enum>>(E);
```
Comment (kuhar) [inline llvm/include/llvm/ADT/STLExtras.h:1542]: do we need the `typename` here?

File: llvm/include/llvm/ADT/STLExtras.h [hunk 2]
```diff
@@ -1535,6 +1535,13 @@ constexpr decltype(auto) makeVisitor(CallableTs &&...Callables) {
   return detail::Visitor<CallableTs...>(std::forward<CallableTs>(Callables)...);
 }
 
+/// Backport of C++23 std::to_underlying.
```
Comment (kuhar) [inline llvm/include/llvm/ADT/STLExtras.h:1538]: Can you write a one-line description of what the function actually does? This being a backport is more of a detail.

**Review by kuhar — COMMENTED** [review]
Makes sense, I saw this implemented in half a dozen of downstream llvm-based projects.

Have you tried grepping the codebase and checking if there are some existing helpers that can be replaced by this?

**Comment by Endilll** [comment]
> Have you tried grepping the codebase and checking if there are some existing helpers that can be replaced by this?

I can't say I performed an exhaustive search, but apparently there is only one helper somewhere deep in MLIR: https://github.com/llvm/llvm-project/blob/0f8615f4dc568f4d7cbf73580eef3e78f64f3bd0/mlir/lib/Dialect/SparseTensor/IR/Detail/TemplateExtras.h#L41-L45

**Comment by kuhar** [comment]
> > Have you tried grepping the codebase and checking if there are some existing helpers that can be replaced by this?
> 
> I can't say I performed an exhaustive search, but apparently there is only one helper somewhere deep in MLIR:
> 
> https://github.com/llvm/llvm-project/blob/0f8615f4dc568f4d7cbf73580eef3e78f64f3bd0/mlir/lib/Dialect/SparseTensor/IR/Detail/TemplateExtras.h#L41-L45

Thanks for checking, @Endilll. Would be nice to replace this with the function from this PR.

File: llvm/include/llvm/ADT/STLExtras.h (hunk 1 above)
Comment (Endilll) [reply llvm/include/llvm/ADT/STLExtras.h:1542]: Apparently we don't need `typename` in both definition and return type: https://godbolt.org/z/nczqn85sn
I'll remove it

File: llvm/include/llvm/ADT/STLExtras.h (hunk 2 above)
Comment (Endilll) [reply llvm/include/llvm/ADT/STLExtras.h]: Done!

File: llvm/include/llvm/ADT/STLExtras.h (hunk 1 above)
Comment (Endilll) [reply llvm/include/llvm/ADT/STLExtras.h]: Done

**Comment by Endilll** [comment]
@slinder1 I didn't know `STLForwardCompat.h` even exists! Thank you for mentioning.

**Comment by Endilll** [comment]
@kuhar Would it be fine to leave such replacement out of scope of this PR?

**Comment by kuhar** [comment]
> @kuhar Would it be fine to leave such replacement out of scope of this PR?

I'd land it together, otherwise this change has no uses in the codebase. And the change is very simple.

File: llvm/unittests/ADT/STLForwardCompatTest.cpp [hunk 3]
```diff
@@ -119,4 +119,20 @@ TEST(TransformTest, MoveTransformLlvm) {
...
+  static_assert(llvm::to_underlying(A1) == 0);
+  static_assert(llvm::to_underlying(B1) == -1);
+
+  enum E2 : unsigned char { A2 = 0, B2 };
+  static_assert(std::is_same_v<unsigned char, decltype(llvm::to_underlying(A2))>);
+  static_assert(llvm::to_underlying(A2) == 0);
+  static_assert(llvm::to_underlying(B2) == 1);
+
+  enum class E3 { A3 = -1, B3 };
+  static_assert(std::is_same_v<int, decltype(llvm::to_underlying(E3::A3))>);
```
Comment (kuhar) [inline llvm/unittests/ADT/STLForwardCompatTest.cpp:133]: Is this guaranteed to be an `int`? Maybe specify the `enum` type to make this clear -- it won't change test coverage.

Comment (kuhar) [reply llvm/unittests/ADT/STLForwardCompatTest.cpp:133]: See https://timsong-cpp.github.io/cppwp/n4659/dcl.enum#7

**Comment by Endilll** [comment]
> I'd land it together, otherwise this change has no uses in the codebase. And the change is very simple.

This is driven by my work on Clang enums, so I intentionally factored this PR out so that this useful facility is not tied to any particular usage (that might get reverted). I'm going to commit several NFC patches tomorrow that use this.

File: llvm/unittests/ADT/STLForwardCompatTest.cpp (hunk 3 above)
Comment (AaronBallman) [reply llvm/unittests/ADT/STLForwardCompatTest.cpp:133]: https://eel.is/c++draft/enum#dcl.enum-5.sentence-4

**Comment by kuhar** [comment]
> > I'd land it together, otherwise this change has no uses in the codebase. And the change is very simple.
> 
> This is driven by my work on Clang enums, so I intentionally factored this PR out so that this useful facility is not tied to any particular usage (that might get reverted). I'm going to commit several NFC patches tomorrow that use this.

The clang part makes sense to me, but the other usage seems to have exactly the same API and implementation and should be completely safe to replace.

File: llvm/unittests/ADT/STLForwardCompatTest.cpp (hunk 3 above)
Comment (Endilll) [reply llvm/unittests/ADT/STLForwardCompatTest.cpp:133]: Clause 5 above states that *For a scoped enumeration type, the underlying type is int if it is not explicitly specified.*

**Comment by AaronBallman** [comment]
> > @kuhar Would it be fine to leave such replacement out of scope of this PR?
> 
> I'd land it together, otherwise this change has no uses in the codebase. And the change is very simple.

FWIW, it's pretty typical for follow-up commits to start using the new ADT functionality; this separation makes it a bit easier to revert if necessary (when the ADT is fine in general but its specific use as a replacement is problematic for some reason).

**Review by kuhar — APPROVED** [review]

**Comment by kuhar** [comment]
> > > @kuhar Would it be fine to leave such replacement out of scope of this PR?
> > 
> > 
> > I'd land it together, otherwise this change has no uses in the codebase. And the change is very simple.
> 
> FWIW, it's pretty typical for follow-up commits to start using the new ADT functionality; this separation makes it a bit easier to revert if necessary (when the ADT is fine in general but its specific use as a replacement is problematic for some reason).

In this specific scenario we would be replacing the same implementation with the same API. Because this is a detail, I don't think it's worth arguing over and I approved the PR as-is.

**Pull request merged**